### ISSUE TITLE
Invert Colors and Make Bitonal glitches

### ISSUE THREAD

**Issue by bricewge** [baskerville/plato#79]
Enabling together "Invert Colors" and "Make Bitonal" glitches, only one half is bitonal and the other half has it's colors inverted. Taking a [screenshot](https://user-images.githubusercontent.com/5525646/70809973-3730e180-1dc3-11ea-98c0-d6c24b3a5b57.png) of it won't show the behaviour, you can see it in [this video](https://0x0.st/zUUy.mp4).

**Comment by baskerville** [comment]
I guess it might have to do with the fact that I'm using the legacy *ioctl* calls and you have a *Mark 7* device. I'll provide a test binary. Side note: you can select a menu entry without closing the menu by taping and holding the entry.

**Comment by baskerville** [comment]
Could you try the attached binary?

[plato-0.8.3-test1.zip](https://github.com/baskerville/plato/files/4215254/plato-0.8.3-test1.zip)

**Comment by bricewge** [comment]
Nope, the glitch is still here.

**Comment by baskerville** [comment]
I'm afraid it might have to do with a firmware [bug](https://github.com/NiLuJe/FBInk/blob/d1aaa20f67a3d7cb216ada05cce9199cae172891/fbink.c#L1990) then. @NiLuJe: Can't `EPDC_FLAG_ENABLE_INVERSION` and `EPDC_FLAG_FORCE_MONOCHROME` work together on the *Forma*?

**Comment by NiLuJe** [comment]
IIRC, it can, but there's a weird bug where the EPDC will sometimes *ignore* any postprocessing flags (any combination of flags and/or dithering_mode).

I've seen weird sporadic issues like that over the years, but it's much worse and somewhat reproducible on Mk. 7 devices, as the comment linked above shows ;).

It's been my experience that bitdepth and/or rotation play a large part in it.

That said, I've never seen it happen quite like the OP ^^.

@baskerville: Do you simply add the FORCE_MONOCHROME flag to whatever waveform mode ought to be used usually, or do you also force A2 or DU?

**Comment by NiLuJe** [comment]
At a very quick glance, I certainly can't reproduce that on Nickel (32bpp, Portrait (-> 3)), at least w/ A2. (And I've never seen it happen on other devices w/ DU, either).

TL;DR: That works:
```
21:39:57 [76ef81f6] ioctl(3, MXCFB_SEND_UPDATE, {update_region={top=0, left=0, width=1440, height=1920}, waveform_mode=NTX_WFM_MODE_A2, update_mode=UPDATE_MODE_PARTIAL, update_marker=2276, temp=TEMP_USE_AMBIENT, flags=EPDC_FLAG_ENABLE_INVERSION|EPDC_FLAG_FORCE_MONOCHROME, dither_mode=EPDC_FLAG_USE_DITHERING_PASSTHROUGH, quant_bit=0, alt_buffer_data={phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x7edaeb08) = 0
```

So does piling dithering on top of it:

```
21:41:46 [76e911f6] ioctl(3, MXCFB_SEND_UPDATE, {update_region={top=0, left=0, width=1440, height=1920}, waveform_mode=NTX_WFM_MODE_A2, update_mode=UPDATE_MODE_PARTIAL, update_marker=2282, temp=TEMP_USE_AMBIENT, flags=EPDC_FLAG_ENABLE_INVERSION, dither_mode=EPDC_FLAG_USE_DITHERING_ORDERED, quant_bit=1, alt_buffer_data={phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x7ee96af8) = 0
```

or

```
21:42:21 [76f291f6] ioctl(3, MXCFB_SEND_UPDATE, {update_region={top=0, left=0, width=1440, height=1920}, waveform_mode=NTX_WFM_MODE_A2, update_mode=UPDATE_MODE_PARTIAL, update_marker=2295, temp=TEMP_USE_AMBIENT, flags=EPDC_FLAG_ENABLE_INVERSION|EPDC_FLAG_USE_DITHERING_Y1, dither_mode=EPDC_FLAG_USE_DITHERING_PASSTHROUGH, quant_bit=0, alt_buffer_data={phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x7e94aae8) = 0
```

**Comment by baskerville** [comment]
> Do you simply add the FORCE_MONOCHROME flag to whatever waveform mode ought to be used usually, or do you also force A2 or DU?

The former.

**Comment by NiLuJe** [comment]
I'd try the latter then, I fully expect FORCE_MONOCHROME to be unreliable on anything that isn't A2 or DU ;).

**Comment by baskerville** [comment]
> I'd try the latter then, I fully expect FORCE_MONOCHROME to be unreliable on anything that isn't A2 or DU ;).

The former works on the *Glo HD* and the *Aura ONE*.

**Comment by NiLuJe** [comment]
Although, on that note, extra fun quirk on Mk.7: A2 (possibly DU, too) is a tad too aggressive, and generally will behave weirdly (as in, not updating the non B&W bits) if the content to refresh isn't bitonal *before* FORCE_MONOCHROME.

**Comment by NiLuJe** [comment]
That possibly explain why some UIs have stopped using FORCE_MONOCHROME altogether, even for button highlights, and/or switched away from A2 to DU (or even AUTO, which will generally prefer DU over A2)).

**Comment by NiLuJe** [comment]
TL;DR: The only combination that would provide reliable behavior on Mk. 7 for a global "Bitonal" switch that could apply to effectively any kind of content appears to be `DU` + `EPDC_FLAG_USE_DITHERING_Y1`. Which actually means no `FORCE_MONOCHROME` in practice, since it'd effectively murder the Y1 dithering ;).

**Comment by NiLuJe** [comment]
Oh, duh. It's half and half in the video because the final refresh is regional because it's a menu popping off.

So, yeah, I can *definitely* trigger the weird crap with postprocessing flags getting dropped by the driver if I set invert + monochrome.
Again, still not 100%, but that certainly works as a potential trigger.

----

Also, fun fact: the mess I mentioned about A2 on Mk.7 above? Turns out it's forcce_monochrome's fault, too ;p.

----

So, yeah, force_monochrome is super broken on Mk.7 ^^.

**Comment by baskerville** [comment]
> Although, on that note, extra fun quirk on Mk.7: A2 is a tad too aggressive (and/or buggy ^^), and generally will behave weirdly (as in, not updating the non B&W bits) if the content to refresh isn't bitonal _before_ FORCE_MONOCHROME.

Is this also true for `DU`?

**Comment by NiLuJe** [comment]
Nope, see my latest edits in my previous messages ;).

**Comment by NiLuJe** [comment]
That said, I *can* break DU if I pile on a bunch of different post-processing flags (FORCE_MONOCHROME, INVERSION, and dithering, via flags or dither_mode).

The straw that breaks the camel back appears to be FORCE_MONOCHROME here, as I can't make it go haywire without it.

**Comment by NiLuJe** [comment]
Note that my final test implies that it's still highly plausible that DU + FORCE_MONOCHROME might still randomly break in fun & unexpected ways ;).

It appears *slightly* harder to break than A2, but the offending party is FORCE_MONOCHROME, not A2 ;).

**Comment by baskerville** [comment]
> Note that my final test implies that it's still highly plausible that DU + FORCE_MONOCHROME might still randomly break in fun & unexpected ways ;).

I've tried `DU` + `Y1`, but unfortunately, it seemed as slow as `AUTO`, and *Make Bitonal* is supposed to be very fast…

**Comment by NiLuJe** [comment]
Yeah, it's entirely done by the CPU, and it involves a scanline-sized alloc, so it does make sense that performance would take a hit :/.

Oh, well :s.

**Comment by NiLuJe** [comment]
Okay, finally tried this out.

Two things:

* MONOCHROME+INVERSION still breaks with DU. I guess there might be timing constraints involved, since I couldn't actually break that in my simple one-by-one full-screen refresh tests.
* In the same vein, DU+MONOCHROME still exhibits some of the weird "nope, not gonna refresh that bit" behavior I was seeing with A2 :/. (for instance, try rotating the library view, it leads to fun results with leftover bits of sideways text ^^).

TL;DR: FORCE_MONOCHROME is effectively unusable for that purpose on Mk. 7.

(The new ioctl setup works fine, though 👍  ;)).

**Comment by NiLuJe** [comment]
@baskerville: On an unrelated note, popping up the menu in the Reader yields a bunch of 2px high refreshes (the vertical menu borders?).

```
[pid  7075] 15:49:06 [75eac1f6] ioctl(3, MXCFB_WAIT_FOR_UPDATE_COMPLETE, {update_marker=163, collision_test=0}, 0x7ef456a0) = 0
[pid  7075] 15:49:06 [75eac1f6] ioctl(3, MXCFB_SEND_UPDATE, {update_region={top=0, left=0, width=1440, height=119}, waveform_mode=WAVEFORM_MODE_AUTO, update_mode=UPDATE_MODE_PARTIAL, update_marker=165, temp=TEMP_USE_AMBIENT, flags=0, dither_mode=EPDC_FLAG_USE_DITHERING_PASSTHROUGH, quant_bit=0, alt_buffer_data={phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x7ef45798) = 0
[pid  7075] 15:49:06 [75eac1f6] ioctl(3, MXCFB_SEND_UPDATE, {update_region={top=119, left=0, width=1440, height=2}, waveform_mode=WAVEFORM_MODE_AUTO, update_mode=UPDATE_MODE_PARTIAL, update_marker=166, temp=TEMP_USE_AMBIENT, flags=0, dither_mode=EPDC_FLAG_USE_DITHERING_PASSTHROUGH, quant_bit=0, alt_buffer_data={phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x7ef45798) = 0
[pid  7075] 15:49:06 [75eac1f6] ioctl(3, MXCFB_WAIT_FOR_UPDATE_COMPLETE, {update_marker=164, collision_test=0}, 0x7ef45720) = 0
[pid  7075] 15:49:06 [75eac1f6] ioctl(3, MXCFB_SEND_UPDATE, {update_region={top=1487, left=0, width=1440, height=2}, waveform_mode=WAVEFORM_MODE_AUTO, update_mode=UPDATE_MODE_PARTIAL, update_marker=167, temp=TEMP_USE_AMBIENT, flags=0, dither_mode=EPDC_FLAG_USE_DITHERING_PASSTHROUGH, quant_bit=0, alt_buffer_data={phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x7ef45798) = 0
[pid  7075] 15:49:06 [75eac1f6] ioctl(3, MXCFB_SEND_UPDATE, {update_region={top=1489, left=0, width=1440, height=310}, waveform_mode=WAVEFORM_MODE_AUTO, update_mode=UPDATE_MODE_PARTIAL, update_marker=168, temp=TEMP_USE_AMBIENT, flags=0, dither_mode=EPDC_FLAG_USE_DITHERING_PASSTHROUGH, quant_bit=0, alt_buffer_data={phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x7ef45798) = 0
[pid  7075] 15:49:06 [75eac1f6] ioctl(3, MXCFB_SEND_UPDATE, {update_region={top=1799, left=0, width=1440, height=2}, waveform_mode=WAVEFORM_MODE_AUTO, update_mode=UPDATE_MODE_PARTIAL, update_marker=169, temp=TEMP_USE_AMBIENT, flags=0, dither_mode=EPDC_FLAG_USE_DITHERING_PASSTHROUGH, quant_bit=0, alt_buffer_data={phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x7ef45798) = 0
[pid  7075] 15:49:06 [75eac1f6] ioctl(3, MXCFB_SEND_UPDATE, {update_region={top=1801, left=0, width=1440, height=119}, waveform_mode=WAVEFORM_MODE_AUTO, update_mode=UPDATE_MODE_PARTIAL, update_marker=170, temp=TEMP_USE_AMBIENT, flags=0, dither_mode=EPDC_FLAG_USE_DITHERING_PASSTHROUGH, quant_bit=0, alt_buffer_data={phys_addr=0, width=0, height=0, alt_update_region={top=0, left=0, width=0, height=0}}}, 0x7ef45798) = 0
```

I *have* met devices which would deadlock on x1 or 1x refresh rects, I'd probably merge those in the menu rects themselves ;).

**Comment by Aszarsha** [comment]
FWIW, 48a588a did not solve the issue, I can still reproduce (on my Libra).
Regarding to this issue, the part of the text that is de-inverted is exactly the height (but full width) of the burger menu, without the top bar which remains inverted.

Furthermore, and it might be a different issue (or not [*]) : if I change the page (epub viewer, with physical buttons without exiting the menu), the inversion is then... inverted (white background, black foreground) ; monochrome is kept (horrendously aliased rendering). The whole screen is de-inverted in that case, menu including.

When that happens (and maybe in other conditions [**]), and I let the device go into sleeping mode, the "sleeping screensaver" is messed up: the text is ok, but the icon does not appear and the book's text (that was at that position) is kept on the display. Maybe because the icon is not monochrome? See [photo](https://user-images.githubusercontent.com/632292/75368116-89ad7480-58c1-11ea-802b-4dd66f611ac4.jpg).

[*] Do you want me to issue another bug report?
[**] This might also warrant another bug report, but it seems related to that one.

**Comment by baskerville** [comment]
I think the best approach here would be to disable the *Make Bitonal* menu entry on mark 7 devices (and disable *Invert Colors* on the *Aura*).

**Comment by Aszarsha** [comment]
What is the use of bitonal mode? To be sure on my Libra it is a bit faster, but not much, and depending on the font used and its size the text can be barely readable.

**Comment by NiLuJe** [comment]
@Aszarsha: * is essentially my first bullet point in the recap above, and ** is the second ;). So nothing new.

----

@baskerville: If you're not happy w/ Y1, probably, yeah.

**Comment by baskerville** [comment]
> What is the use of bitonal mode? 

*Invert Colors* and *Make Bitonal* were added very early on as a mean to access the corresponding hardware flags, without worrying about their usefulness.

I almost never use *Invert Colors*. On the devices where *Make Bitonal* works, it can be used to skim through pages.

**Comment by Aszarsha** [comment]
> it can be used to skim through pages.

Well on the Libra is certainly is too slow for that purpose, it is barely any faster. Maybe that is why Kobo implemented their skimming interface, which is quite nice.
